tests/Doctrine/Tests/ORM/UnitOfWorkTest.php

```diff
@@ -707,6 +711,48 @@ public function testCanInstantiateInternalPhpClassSubclassFromUnserializedMetada
 
         self::assertInstanceOf(MyArrayObjectEntity::class, $this->unitOfWork->newInstance($classMetadata));
     }
+
+    public function testCallCommitInsideCommit() : void
+    {
+        $listener = new class()
+        {
+            public function preFlush(PreFlushEventArgs $eventArgs)
+            {
+                $eventArgs->getEntityManager()->getUnitOfWork()->commit();
+            }
+        };
+
+        $this->expectException(CommitInsideCommit::class);
+        $this->eventManager->addEventListener(Events::preFlush, $listener);
+        $this->unitOfWork->commit();
+    }
+
+    public function testCommitWithExceptionClearsInProgressFlag() : void
+    {
+        $listener = new class()
+        {
+            public function preFlush(PreFlushEventArgs $eventArgs)
+            {
+                throw new RuntimeException();
+            }
+        };
+
+        $this->eventManager->addEventListener(Events::preFlush, $listener);
+        try {
+            $this->unitOfWork->commit();
+        } catch (RuntimeException $e) {
+        }
+
+        $this->eventManager->removeEventListener(Events::preFlush, $listener);
+
+        try {
+            $this->unitOfWork->commit();
+        } catch (CommitInsideCommit $e) {
+            $this->fail('UnitOfWork::$commitInProgress flag must be reset in case of exception during commit process');
+        }
+
+        $this->assertTrue(true);
```

Comment:
Interesting thread about this: https://github.com/sebastianbergmann/phpunit-documentation/issues/171

Reply:
A better approach to this would be to expect the listener to be called a certain number of times: that would remove the need for the assertion

Reply:
I'll add a counter inside the listener class and check it value here.

Reply:
I've added listener call counter and check, but it looks overcomplicated. Not sure that this is better than solution from sebastianbergmann/phpunit-documentation#171 or `assertTrue(true)`

https://github.com/doctrine/doctrine2/pull/7448/commits/0b3427bcc245c468f48878a0d70dd3c0a11069c9